File: src/core/stdlib/maths/rounding.ts
/**
 * Rounding, and the one rule that decides every case.
 *
 * **Halves go away from zero, never to even.** `stdlib.md` section 8.1 fixes it
 * there and the reason is legibility: a price rounded for display should agree
 * with what a trader would write down, and round-half-to-even surprises people
 * at exactly the values that matter. Fixing it also means two engines cannot
 * differ by one tick.
 *
 * The implementation does not add a half and take the floor. That is the usual
 * shortcut and it is wrong for the value just below a half, where adding 0.5
 * rounds up to the next representable number before the floor ever runs. The
 * fractional part is compared instead, which has no such case.
 */
import type { Value } from '../values/index.js';
import { NONE, isPresent, result } from '../values/index.js';

/** The nearest whole number, halves away from zero. */
export function roundHalfAway(x: number): number {
  const below = Math.floor(x);
  const fraction = x - below;
  if (fraction > 0.5) return below + 1;
  if (fraction < 0.5) return below;
  // Exactly a half. Away from zero: upward above zero, and `below` already is
  // the downward answer for a negative, since floor(-2.5) is -3.
  return x > 0 ? below + 1 : below;
}

/** `round(x)`: to the nearest whole number, halves away from zero. */
export function round(x: Value): Value {
  return isPresent(x) ? result(roundHalfAway(x)) : NONE;
}

/**
 * Ten to each whole power a binary64 can hold, as the binary64 nearest to it.
 *
 * `stdlib.md` 20.7: the scale a fixed decimal rounding or conversion multiplies
 * by is the nearest binary64 to the power of ten, the value the literal `1e23`
 * reads as, and not what a floating point power returns, which is an ulp away
 * from it for one count on this engine's host. Built once from exact integer
 * arithmetic, and it is the one table: `text(x, decimals)` in the engine's
 * library scales by it through `scaleOf` rather than holding a copy.
 */
const POWERS_OF_TEN: readonly number[] = Array.from({ length: 309 }, (_, count) =>
  Number(10n ** BigInt(count)),
);

/**
 * The scale for a digit count: the nearest binary64 to ten to that power, and
 * past the last finite power the infinity the power would have been.
 */
export function scaleOf(decimals: number): number {
  return POWERS_OF_TEN[decimals] ?? Infinity;
}

/**
 * `round(x, decimals)`: to a fixed number of decimals, halves away from zero.
 *
 * A digit count is a whole number of zero or more; anything else is OS3004 or
 * OS4003 before the call reaches here, and absence is the backstop.
 */
export function roundTo(x: Value, decimals: number): Value {
  if (!isPresent(x)) return NONE;
  if (!Number.isInteger(decimals) || decimals < 0) return NONE;
  const scale = scaleOf(decimals);
  return result(roundHalfAway(x * scale) / scale);
}

/** `floor(x)`: toward negative infinity. */
export function floor(x: Value): Value {
  return isPresent(x) ? result(Math.floor(x)) : NONE;
}

/** `ceil(x)`: toward positive infinity. */
export function ceil(x: Value): Value {
  return isPresent(x) ? result(Math.ceil(x)) : NONE;
}

/** `trunc(x)`: toward zero. */
export function trunc(x: Value): Value {
  return isPresent(x) ? result(Math.trunc(x)) : NONE;
}

/** `roundToStep(x, step)`: to the nearest multiple of `step`. */
export function roundToStep(x: Value, step: Value): Value {
  if (!isPresent(x) || !isPresent(step) || step <= 0) return NONE;
  return result(roundHalfAway(x / step) * step);
}

/**
 * `roundToTick(price)`: to the instrument's tick.
 *
 * Absent when the host supplied no tick size, rather than the price unrounded.
 * Returning the input would produce an order price that looks rounded and is
 * not, which is a defect nothing downstream can see.
 */
export function roundToTick(price: Value, tickSize: Value): Value {
  if (!isPresent(tickSize) || tickSize <= 0) return NONE;
  return roundToStep(price, tickSize);
}

File: engine/openscript/library/rounding.py
"""Rounding, and the one rule that decides every case.

**Halves go away from zero, never to even.** `stdlib.md` section 8.1 fixes it
and gives the reason: a price rounded for display should agree with what a
trader would write down, and round half to even surprises people at exactly the
values that matter. Fixing it also means two engines cannot differ by one tick.

**Adding a half and taking the floor is not this function.** It is the usual
shortcut and it is wrong for the value just below a half, where adding 0.5 rounds
up to the next representable number before the floor ever runs and the answer
comes out one too high. `stdlib.md` section 20.7 writes the comparison of the
fractional part out instead, which has no such case, and that is what is below.

**The scale of a fixed decimal rounding is not a floating point power.** 20.7
again: it is the binary64 nearest to ten to that count, the value the literal
``1e23`` reads as. On one host the two differ by an ulp at a count of 23 out of
the 309 a binary64 can hold, and over a price walk that one count moved 814
results. The table here is built once from exact whole number arithmetic, which
is the "exact integer power converted once" 20.7 names, and it is the one table:
the display conversion of ``text(x, decimals)`` scales by it as well.
"""

import math

from .values import ABSENT, Value, floor_of, number, result, whole


def round_half_away(x: float) -> float:
    """The nearest whole number, halves away from zero.

    A value with no whole number near it, an infinity produced by a scaling that
    overflowed, is returned unchanged for ``result`` to turn into absence.
    """
    if not math.isfinite(x):
        return x
    below = floor_of(x)
    fraction = x - below
    if fraction > 0.5:
        return below + 1.0
    if fraction < 0.5:
        return below
    # Exactly a half. Away from zero: upward above zero, and `below` already is
    # the downward answer for a negative, since the floor of -2.5 is -3.
    return below + 1.0 if x > 0 else below


# Ten to each whole power a binary64 can hold, as the binary64 nearest to it.
# Built from exact whole number arithmetic rather than from a floating point
# power, for the reason at the top of this file.
_POWERS_OF_TEN: tuple[float, ...] = tuple(float(10**count) for count in range(309))


def scale_of(decimals: int) -> float:
    """The scale for a digit count: the nearest binary64 to ten to that power.

    Past the last finite power it is the infinity the power would have been, so
    that a count a script computed scales to absence rather than to a wrong
    number.
    """
    if 0 <= decimals < len(_POWERS_OF_TEN):
        return _POWERS_OF_TEN[decimals]
    return math.inf


def floor(x: Value) -> Value:
    """``floor(x)``: toward negative infinity."""
    value = number(x)
    return ABSENT if value is None else result(floor_of(value))


def ceil(x: Value) -> Value:
    """``ceil(x)``: toward positive infinity."""
    value = number(x)
    if value is None:
        return ABSENT
    return result(value if not math.isfinite(value) else float(math.ceil(value)))


def trunc(x: Value) -> Value:
    """``trunc(x)``: toward zero."""
    value = number(x)
    if value is None:
        return ABSENT
    return result(value if not math.isfinite(value) else float(math.trunc(value)))


def round_to_whole(x: Value) -> Value:
    """``round(x)``: to the nearest whole number, halves away from zero."""
    value = number(x)
    return ABSENT if value is None else result(round_half_away(value))


def round_to(x: Value, decimals: Value) -> Value:
    """``round(x, decimals)``: scale, round once, scale back."""
    value = number(x)
    count = whole(decimals)
    if value is None or count is None:
        return ABSENT
    scale = scale_of(count)
    return result(round_half_away(value * scale) / scale)


def round_to_step(x: Value, step: Value) -> Value:
    """``roundToStep(x, step)``: to the nearest multiple of ``step``."""
    value = number(x)
    size = number(step)
    if value is None or size is None or size <= 0:
        return ABSENT
    return result(round_half_away(value / size) * size)


def round_to_tick(price: Value, tick_size: Value) -> Value:
    """``roundToTick(price)``: to the instrument's tick.

    Absent when the host has stated no tick size, rather than the price
    unrounded. Returning the input would produce an order price that looks
    rounded and is not, and a venue refuses an order off the tick, so a script
    that cannot round has to be able to see that it cannot.
    """
    tick = number(tick_size)
    if tick is None or tick <= 0:
        return ABSENT
    return round_to_step(price, tick)
